File: wiz-baseball-game-java/src/main/java/BaseballGame.java
import static helper.InputHelper.input;

import java.util.*;
import java.util.stream.Collectors;

public class BaseballGame {
	private final List<Character> availableChars;
	private final int size;
	private final String answer;

	private int tryCount = 0;

	public BaseballGame() {
		this(3, List.of('1', '2', '3', '4', '5', '6', '7', '8', '9'));
	}

	public BaseballGame(int size) {
		this(size, List.of('1', '2', '3', '4', '5', '6', '7', '8', '9'));
	}

	public BaseballGame(List<Character> availableChars) {
		this(3, availableChars);
	}

	public BaseballGame(int size, List<Character> availableChars) {
		this.size = size;
		this.availableChars = availableChars;
		this.answer = this.generateAnswer();
	}

	/**
	 * [ BaseballGame.play() ]
	 * 1) 사용자로부터 정답을 입력받는다.
	 * 2) 사용자의 입력이 올바른지 검증한다.  -> 추상화 수준에 대한 설명 필수!!!
	 * 3) tryCount 를 1 증가시킨다.
	 * 4) 스트라이크 개수를 센다.
	 * 5) 스트라이크 개수가 정답의 길이와 같다면, 정답으로 처리하고 게임을 종료한다.
	 * 6) 볼 개수를 센다.
	 * 7) 힌트를 노출하고 게임을 계속한다.
	 */
	public void play() {
		System.out.println("\n< 게임을 시작합니다 >");

		while (true) {
			String input = input("정답을 입력하세요 : ");
			if (!this.validateInput(input)) {
				System.out.println("잘못된 입력입니다. 다시 입력해주세요!");
				continue;
			}

			this.tryCount++;
			int strike = this.countStrike(input);
			if (strike == this.size) {
				System.out.println(strike + " 스트라이크! 정답입니다!");
				break;
			}
			int ball = this.countBall(input);
			System.out.println(strike + " 스트라이크, " + ball + " 볼!");
		}
	}

	public String getResult() {
		return "시도 횟수 - " + this.tryCount;
	}

	private boolean validateInput(String input) {
		Set<Character> seen = new HashSet<>();
		for (char c : input.toCharArray()) {
			if (!seen.add(c))
				return false;
			if (!availableChars.contains(c))
				return false;
		}
		return seen.size() == this.size;
	}

	/**
	 * generateAnswer() 와 같은건 작은 문제!!
	 * 전체적인 이미지와 흐름을 결정하는게 먼저고, 그 과정에서 분할된 작은 문제들이 나온다.
	 * "관심사의 분리" 를 통해서 작은 문제들을 더욱 쉽게 해결할 수 있고, 이 전체적인 과정이 일종의 CS 관점에서 분할정복이라고 할 수 있다.
	 * 우리는 작은 문제에만 집중해서 더욱 문제를 쉽고 빠르게 해결할 수 있다.
	 */
	private String generateAnswer() {
		List<Character> targets = new ArrayList<>(availableChars);
		Collections.shuffle(targets);
		return targets.stream()
			.limit(this.size)
			.map(String::valueOf)
			.collect(Collectors.joining());
	}

	private int countStrike(String input) {
		var count = 0;
		for (int i = 0; i < input.length(); i++) {
			if (input.charAt(i) == this.answer.charAt(i)) {
				count++;
			}
		}
		return count;
	}

	private int countBall(String input) {
		var count = 0;
		for (int i = 0; i < input.length(); i++) {
			if (input.charAt(i) != this.answer.charAt(i)
				&& this.answer.contains(String.valueOf(input.charAt(i)))) {
				count++;
			}
		}
		return count;
	}
}

File: wiz-baseball-game/src/main/kotlin/BaseballGame.kt
import helper.input

open class BaseballGame(private val size: Int = 3) {

    protected open val answer = (1..9).shuffled().take(size)

    /**
     *   BaseballGame.play()
     *    1. tryCount 을 0 으로 초기화한다.
     *    2. 사용자에게 정답 입력을 받는다.
     *    3. 사용자에게 입력받은 입력이 올바른지 검증한다.
     *    4. tryCount 을 1 증가시킨다.
     *    5. 스트라이크가 몇개인지 계산한다.
     *    6. 만약 스트라이크가 size 와 동일하다면 정답임으로 게임을 종료한다. (이때 TryCount 반환한다)
     *    7. 볼이 몇개인지 계산한다.
     *    8. 힌트를 출력하고 처음으로 돌아간다.
     *
     *   // 적당한 추상화가 적용된 설명 -> 읽기 쉬운 코드 만들어짐
     *   // 캡슐화를 구현할 수 있다 -> 구체적인걸 늦게 결정하기 때문
     *   // 관심사 분리를 통해서 문제를 작게 만들고, 생각하는 범위자체를 작게 만들수 있다.
     */
    fun play(): Int {
        println("\n< 게임을 시작합니다 >")
        var tryCount = 0

        while (true) {
            // TODO : 사용자 입력을 별도로 함수를 통해서 받을지 말지는 선택의 문제다!
            val input = try {
                input("정답을 입력하세요 : ")
                    .toList()
                    .map { it.digitToInt() }
                    .also { if (!this.checkInput(it)) throw RuntimeException() }
            } catch (e: RuntimeException) {
                println("잘못된 입력입니다. 다시 입력해주세요!")
                continue
            }

            tryCount++
            val strike = this.countStrike(input)
            if (strike == size) {
                println("$strike 스트라이크! 정답입니다!")
                break
            }
            val ball = this.countBall(input)
            println("$strike 스트라이크! $ball 볼!")
        }

        println()
        return tryCount
    }

    protected open fun checkInput(input: List<Int>): Boolean {
        if (input.size != size) return false
        if (input.distinct().size != size) return false
        return true
    }

    private fun countStrike(input: List<Int>): Int {
        var count = 0
        for (i in 0..<size) {
            if (input[i] == answer[i]) {
                count++
            }
        }
        return count
    }

    private fun countBall(input: List<Int>): Int {
        var count = 0
        for (i in 0..<size) {
            if (input[i] != answer[i] && input[i] in answer) {
                count++
            }
        }
        return count
    }
}